src/foremast/datapipeline/datapipeline.py

```diff
@@ -0,0 +1,109 @@
+        self.env = env
+        self.region = region
+        self.properties = get_properties(prop_path)
+        self.datapipeline_data = self.properties[self.env]['datapipeline']
+        generated = get_details(app=self.app_name)
+        self.group = generated.data['project']
+
+        session = boto3.Session(profile_name=self.env, region_name=self.region)
+        self.client = session.client('datapipeline')
+        self.pipeline_id = None
+
+    def create_datapipeline(self):
+        """Creates the data pipeline if it does not already exist"""
+
+        tags = [
+                {"key": "app_group",
+                "value": self.group},
+                {"key": "app_name",
+                "value": self.app_name}
+                ]
+        response = self.client.create_pipeline(name=self.datapipeline_data.get('name', self.app_name),
+                                               uniqueId=self.app_name,
+                                               description=self.datapipeline_data['description'],
+                                               tags=tags)
+        self.pipeline_id = response.get('pipelineId')
+        LOG.debug(response)
+        LOG.info("Successfully configured Data Pipeline - %s", self.app_name)
+
+    def set_pipeline_definition(self):
+        """Translates the json definition and puts it on created pipeline"""
+
+        if not self.pipeline_id:
+            self.get_pipeline_id()
+
+        pipelineObjects = translator.definition_to_api_objects(self.datapipeline_data['json_definition'])
+        parameterObjects = translator.definition_to_api_parameters(self.datapipeline_data['json_definition'])
+        parameterValues = translator.definition_to_parameter_values(self.datapipeline_data['json_definition'])
+        response = self.client.put_pipeline_definition(
+                                pipelineId=self.pipeline_id,
+                                pipelineObjects=pipelineObjects,
+                                parameterObjects=parameterObjects,
+                                parameterValues=parameterValues)
+        LOG.debug(response)
+        LOG.info("Successfully applied pipeline definition")
+
+    def get_pipeline_id(self):
+        """Finds the pipeline ID for configured pipeline"""
+
+        all_pipelines = []
+        hasMore = True
+        marker = None
+        # handles the pagination from boto3
+        while hasMore:
+            if not marker:
+                response = self.client.list_pipelines()
+            else:
+                LOG.info("Checking for more pipelines")
+                response = self.client.list_pipelines(marker=marker)
+            all_pipelines.extend(response['pipelineIdList'])
+            hasMore = response['hasMoreResults']
+            marker = response.get('marker')
+            LOG.debug(response)
+        
+        for pipeline in all_pipelines:
+            if pipeline['name'] == self.datapipeline_data.get('name', self.app_name):
```

Comment:
I think it would be better. We would get an O(1) lookup against a dictionary in pretty much all scenarios. The single `.get()` operation also tells me quickly that we either get the ID, or `None`, in one place without having to look at the `__init__()`.

If we also want to use Cyclomatic Complexity Number as an objective measure of complexity, the requested changes would reduce the CCN by one unit.

Before:
```bash
(foremast) ntangsurat@derigible ~/git/github.foremast feature/data_pipeline $ lizard src/foremast/datapipeline/datapipeline.py
================================================
  NLOC    CCN   token  PARAM  length  location
------------------------------------------------
      11      1    113      5      20 __init__@30-49@src/foremast/datapipeline/datapipeline.py
      14      1    101      1      16 create_datapipeline@51-66@src/foremast/datapipeline/datapipeline.py
      14      2     87      1      17 set_pipeline_definition@68-84@src/foremast/datapipeline/datapipeline.py
      11      4     87      1      14 get_pipeline_id@86-99@src/foremast/datapipeline/datapipeline.py
1 file analyzed.
==============================================================
NLOC    Avg.NLOC  AvgCCN  Avg.token  function_cnt    file
--------------------------------------------------------------
     57      12.5     2.0       97.0         4     src/foremast/datapipeline/datapipeline.py

=============================================================================================
No thresholds exceeded (cyclomatic_complexity > 15 or length > 1000 or parameter_count > 100)
==========================================================================================
Total nloc   Avg.NLOC  AvgCCN  Avg.token   Fun Cnt  Warning cnt   Fun Rt   nloc Rt
------------------------------------------------------------------------------------------
        57      12.5     2.0       97.0        4            0      0.00    0.00
```

After:
```bash
(foremast) ntangsurat@derigible ~/git/github.foremast feature/data_pipeline $ lizard src/foremast/datapipeline/datapipeline.py
================================================
  NLOC    CCN   token  PARAM  length  location
------------------------------------------------
      11      1    113      5      20 __init__@30-49@src/foremast/datapipeline/datapipeline.py
      10      1    101      1      12 create_datapipeline@51-62@src/foremast/datapipeline/datapipeline.py
      14      2     87      1      17 set_pipeline_definition@64-80@src/foremast/datapipeline/datapipeline.py
       8      3     68      1      11 get_pipeline_id@82-92@src/foremast/datapipeline/datapipeline.py
1 file analyzed.
==============================================================
NLOC    Avg.NLOC  AvgCCN  Avg.token  function_cnt    file
--------------------------------------------------------------
     50      10.8     1.8       92.2         4     src/foremast/datapipeline/datapipeline.py

=============================================================================================
No thresholds exceeded (cyclomatic_complexity > 15 or length > 1000 or parameter_count > 100)
==========================================================================================
Total nloc   Avg.NLOC  AvgCCN  Avg.token   Fun Cnt  Warning cnt   Fun Rt   nloc Rt
------------------------------------------------------------------------------------------
        50      10.8     1.8       92.2        4            0      0.00    0.00
```

Neither code blocks will handle the strange case where we have duplicate Data Pipeline names. Also, using `laboratory`, it looks like both current and proposed code runs in similar times with a small data set.

```bash
(foremast) ntangsurat@derigible ~/git/github.foremast feature/data_pipeline $ ./t.py
Current result: 0:00:00.170039
Proposed result:        0:00:00.158559
Current result: 0:00:00.166175
Proposed result:        0:00:00.152733
Current result: 0:00:00.189293
Proposed result:        0:00:00.195909
Current result: 0:00:00.186727
Proposed result:        0:00:00.175716
Current result: 0:00:00.171098
Proposed result:        0:00:00.167694
Current result: 0:00:00.171337
Proposed result:        0:00:00.180104
Current result: 0:00:00.183483
Proposed result:        0:00:00.174779
Current result: 0:00:00.188255
Proposed result:        0:00:00.177035
Current result: 0:00:00.177887
Proposed result:        0:00:00.170697
Current result: 0:00:00.173281
Proposed result:        0:00:00.168539
```

I'm willing to drop this issue.